### PR TITLE
rewrite Pycell documentation

### PR DESCRIPTION
**Pull request by mejrs**
Reorganizes and adds a bit of documentation.

@kngwyu you wrote much of what was previously there, would you be willing to review?

Rendered:

afbeelding
afbeelding
afbeelding

afbeelding

### CODE REVIEW

File: src/pycell.rs [hunk 1]
```diff
@@ -1,6 +1,144 @@
...
+//! #
+//! # #[pymethods]
+//! # impl Number {
+//! #     fn increment(&mut self) {
+//! #         self.inner += 1;
+//! #     }
+//! # }
+//! # fn main() -> PyResult<()> {
+//! Python::with_gil(|py| {
+//!     let n = PyCell::new(py, Number { inner: 0 })?;
```
Comment (birkenfeld) [inline src/pycell.rs:76]: Can we start from a `Py<Number>` or instead? That is much more common that constructing a `PyCell` ourselves.

File: src/pycell.rs [hunk 2]
```diff
@@ -1,6 +1,144 @@
...
+//! It is also necessary to use [`PyCell`] if you can receive mutable arguments that may overlap.
+//! Suppose the following function that swaps the values of two `Number`s:
+//! ```
+//! # use pyo3::prelude::*;
+//! # #[pyclass]
+//! # pub struct Number {
+//! #     inner: u32,
+//! # }
+//! #[pyfunction]
+//! fn swap_numbers(a: &mut Number, b: &mut Number){
```
Comment (birkenfeld) [inline src/pycell.rs:107]:
```suggestion
//! fn swap_numbers(a: &mut Number, b: &mut Number) {
```

File: src/pycell.rs [hunk 3]
```diff
@@ -1,6 +1,144 @@
...
+//!     std::mem::swap(&mut a.inner, &mut b.inner);
+//! }
+//! ```
+//! When users pass in the same `Number` as both arguments, one of the mutable borrows will
+//! fail and raise a `RuntimeError`:
+//! ```text
+//! >>> a = Number()
+//! >>> swap_numbers(a,a)
```
Comment (birkenfeld) [inline src/pycell.rs:115]:
```suggestion
//! >>> swap_numbers(a, a)
```

File: src/pycell.rs [hunk 4]
```diff
@@ -32,71 +170,41 @@ pub struct PyCellBase<T> {
...
 ///     }
 /// }
-/// # Python::with_gil(|py| {
-/// #     let counter = PyCell::new(py, Counter::default()).unwrap();
-/// #     pyo3::py_run!(py, counter, "assert counter.increment('cat') == 1");
-/// # });
+///
+/// # fn main() -> PyResult<()> {
+/// Python::with_gil(|py| {
+///     let n = PyCell::new(py, Number { inner: 0 })?;
```
Comment (birkenfeld) [inline src/pycell.rs:197]: Here the example using `PyCell::new` is fine I think.

File: src/pycell.rs (hunk 1 above)
Comment (mejrs) [reply src/pycell.rs]: Done :)

File: src/pycell.rs [hunk 5]
```diff
@@ -77,7 +77,7 @@
 //!
 //!     // We borrow the guard and then dereference
 //!     // it to get a mutable reference to Number
-//!     let mut guard: PyRefMut<'_, Number> = n.borrow_mut();
+//!     let mut guard: PyRefMut<'_, Number> = n.as_ref(py).borrow_mut();
```
Comment (birkenfeld) [inline src/pycell.rs:80]: Ah I think `n.borrow_mut(py)` should also work. But it probably doesn't matter.

**Review by birkenfeld — APPROVED** [review]

File: src/pycell.rs [hunk 6]
```diff
@@ -568,16 +683,15 @@ impl<T: PyClass + fmt::Debug> fmt::Debug for PyRef<'_, T> {
     }
 }
 
-/// Wraps a mutable borrowed reference to a value in a `PyCell<T>`.
+/// A wrapper type for a mutably borrowed value from a[`PyCell`]`<T>`.
 ///
-/// See the [`PyCell`](struct.PyCell.html) and [`PyRef`](struct.PyRef.html) documentation for more.
+/// See the [module-level documentation](self) for more information.
```
Comment (davidhewitt) [inline src/pycell.rs:688]: Haha nice I didn't know this works 😄

**Review by davidhewitt — COMMENTED** [review]
Looks really good to me - do you want to wait for @kngwyu to review or shall we proceed to merge? I understand they have now started their PhD and wanted to more or less step back from OSS for now to focus on that.

**Comment by mejrs** [comment]
I was unaware that he is busy with other things now. Let's just merge.

**Comment by kngwyu** [comment]
Thanks, I like the `swap` example.

> I understand they have now started their PhD and wanted to more or less step back from OSS for now to focus on that.

Yeah I hope I can do some contribution again at some time, but I'm quite busy for now...

**Pull request merged**